1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py

```diff
@@ -0,0 +1,320 @@
+mlflow.log_params(optim_params)
+
+
+# In[ ]:
+
+
+# These keys will be in the names of the logged losses
+tracked_losses = {
+    "mse loss": nn.MSELoss(),
+    "mae loss": nn.L1Loss()
+}
+
+backprop_loss_name = "mae loss"
+
+mlflow.log_param("Training Loss", backprop_loss_name)
+
+
+# In[ ]:
+
+
+trainer_params = {
+    "_batch_size": 32,
+    "_epochs": 2_000,
+    "_patience": 15
+}
+
+
+# In[ ]:
+
+
+trainer = ModelTrainer(
+    _model=model,
+    _image_dataset=img_dataset,
+    _optimizer=optimizer,
+    _tracked_losses=tracked_losses,
+    _backprop_loss_name=backprop_loss_name,
+    **trainer_params
+)
+
+
+# In[ ]:
+
+
+trainer.train()
+
+
+# # Generate Images
+# Evaluate the model by generating the same number of example images for each siRNA.
+
+# In[2]:
+
+
+example_images_per_sirna = 10
+max_pixel_val = 2**16 - 1
+
+sirna_img_counts = {sirna: 0 for sirna in scdfs.loc[scdfs["Metadata_Condition"] != "untreated"]["Metadata_Condition"].unique()}
+i = 0
+
+for input, target in iter(trainer.val_dataset):
+
+    img_name = trainer.val_dataset.dataset.input_name
+    img_name = img_name.replace("_illumcorrect.tiff", "")
+
+    cell_id = img_name.split("_")[0]
+    sirna = scdfs.loc[int(cell_id)]["Metadata_Condition"]
+
+    if sirna not in list(sirna_img_counts.keys()):
+        continue
+
+    input, target, output = evaluate_and_format_imgs(input, target)
+
+    titles = ['DAPI Image', 'Predicted GOLD Image', 'Target GOLD Image']
+
+    fig, axes = plt.subplots(1, 3, figsize=(15, 5))
+
+    fig.suptitle(f"{img_name}", fontsize=16)
+
+    plt.subplots_adjust(wspace=0.3, hspace=0)  # Adjust `wspace` if titles overlap
+
+    axes[0].imshow(input, cmap="grey", vmin=0, vmax=max_pixel_val)
+    axes[0].axis('off')
+    axes[0].set_title(titles[0], fontsize=14)
+
+    axes[1].imshow(output, cmap="grey", vmin=0, vmax=max_pixel_val)
+    axes[1].axis('off')
+    axes[1].set_title(titles[1], fontsize=14)
+
+    axes[2].imshow(target, cmap="grey", vmin=0, vmax=max_pixel_val)
+    axes[2].axis('off')
+    axes[2].set_title(titles[2], fontsize=14)
+
+    plt.savefig(figure_path / f"{img_name}.png", bbox_inches='tight', pad_inches=0)
+    plt.close()
+
+    sirna_img_counts[sirna] += 1
+
+    if sirna_img_counts[sirna] >= example_images_per_sirna:
+        sirna_img_counts.pop(sirna)
+
+        if not sirna_img_counts:
+            break
+
+
+# # Log Metrics and Model
+
+# In[ ]:
+
+
+client = mlflow.MlflowClient()
+
+run_id = mlflow.active_run().info.run_id
+run = client.get_run(run_id)
+
+metrics_per_epoch = defaultdict(list)
+
+for metric_name in run.data.metrics.keys():
+    metric_history = client.get_metric_history(run_id=run_id, key=metric_name)
+    for metric in metric_history:
+        metrics_per_epoch[metric_name].append(metric.value)
+
+metricsdf = pd.DataFrame(metrics_per_epoch)
+metricsdf["epoch"] = np.arange(metricsdf.shape[0])
+metricsdf.to_csv(metrics_path / "fnet_metrics_per_epoch.csv", index=False)
```

Comment:
consider writing to parquet or database?

Reply:
Could be useful, Greg wanted me to store it in a tidy csv format for now